### ISSUE TITLE
JSON: Allow weakly-typed mapping via Dictionary

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#29825]
#28871 tracks weakly-typed JSON mapping via JsonDocument/JsonElement, where the JSON document schema varies and so a strongly-typed model isn't appropriate.

We can also allow simply mapping arbitrary Dictionary types - this corresponds to how many people use 

* This intersects with [#28688](https://github.com/dotnet/efcore/issues/28688) (primitive collections in JSON).
* Both keys and values should support any valid JSON data type (string, int...).
* However we could also support `Dictionary<string, object>`, to map an entire hierarchy. See [#26903](https://github.com/dotnet/efcore/issues/26903) on this (currently seemed blocked by property bag detection).
* This should be supported at the top-level, so that the JSON object as a whole is mapped as a single dictionary (see e.g. [this issue](https://github.com/dotnet/efcore/issues/26903#issuecomment-1345392217)). This would mean supporting `OwnsOne(x => x.Json, builder => { builder.ToJson(); })` where x.Json is a Dictionary.

**Comment by roji** [comment]
Duplicate of #29427

**Comment by ajcvickers** [comment]
Re-opening, since dictionary mapping was not covered by "primitive collections".

**Comment by vchirikov** [comment]
Link to the "issue" should be in the docs at least, was difficult to find..
Is it planned for 9.0 release?

**Comment by roji** [comment]
@vchirikov the milestone generally expresses whether an issue is planned for a given release or not.

**Comment by roji** [comment]
@onionhammer that would be a combination of this and #28871.

**Comment by onionhammer** [comment]
@roji JsonElement can already be saved/serialized as an entity property, but Dictionary<string, JsonElement> cannot

**Comment by bugproof** [comment]
is this why this isn't working ? 

```cs
modelBuilder.Entity<Account>()
            .OwnsOne(c => c.Settings, d =>
            {
                d.ToJson();
                d.OwnsOne(a => a.ShippingRateMarkups);
            }).HasKey(c => c.AccountId);
            
public class Settings
{
    public Dictionary<string, decimal>? ShippingRateMarkups { get; set; }
}
```

**Comment by roji** [comment]
@bugproof supporting mapping `Dictionary<string, decimal>` to JSON is what this issue tracked.

As a workaround, if all you want to do is save and load it to the database, you can use a value converter to do the JSON serialization/deserialization yourself. But this won't support querying into it, etc.

**Comment by bugproof** [comment]
And how will it be stored? As JSON instead of JSONB? Yes I don't need complex queries with it, just load/save as settings dictionary isn't very complex.

**Comment by roji** [comment]
> As JSON instead of JSONB?

It's up to you to configure that database column type, as always - [see the docs](https://learn.microsoft.com/en-us/ef/core/modeling/entity-properties?tabs=data-annotations%2Cwithout-nrt#column-data-types).

**Comment by aleksvujic** [comment]
We are using .NET 9, EF Core 9 and Npgsql 9. We are using database-first approach.

We have the following table in the PostgreSQL database:

```sql
CREATE TABLE fit3.results (
	id serial4 NOT NULL,
	"output" jsonb NULL,
	CONSTRAINT results_pk PRIMARY KEY (id)
);
```

It contains the following data:

| id | output |
| -- | -------- |
| 1 |     |
| 2 | {}     |
| 3 | {"hello": "world"}    |
| 4 | {"john": "doe", "hello": "world"}    |

It is scaffolded into the following C# model:

```cs
[Table("results", Schema = "fit3")]
public partial class Result
{
    [Key]
    [Column("id")]
    public int Id { get; set; }

    [Column("output", TypeName = "jsonb")]
    public string? Output { get; set; }
}
```

We created a custom partial class:

```cs
public partial class Result
{
    public Output? OutputJSON { get; set; }
}

public class Output
{
    [JsonPropertyName("data")]
    public Dictionary<string, string?>? Data { get; set; }
}
```

This is entity configuration in `OnModelCreatingPartial`:

```cs
modelBuilder.Entity<Result>(entity =>
{
    entity.OwnsOne(x => x.OutputJSON, x =>
    {
        x.ToJson("output");
        
        // option 1:
        x.OwnsOne(y => y.Data);
        
        // option 2:
        //x.Property(y => y.Data)
        //    .HasConversion(
        //        v => JsonSerializer.Serialize(v, (JsonSerializerOptions?)null),
        //        v => JsonSerializer.Deserialize<Dictionary<string, string?>?>(v, (JsonSerializerOptions?)null)
        //    );
    });
});
```

We try to perform the simplest query:

```cs
var results = fitDbContext.Results.ToList();
```

We tried both options (option 1 with `OwnsOne` and option 2 with `HasConversion`), but the `Result.OutputJSON.Data` field is always `null`. What are we doing wrong?

**Comment by roji** [comment]
@aleksvujic EF doesn't (yet) support mapping a Dictionary to JSON - that's what this issue tracks. If the actual contents of your JSON column (`output`) has a stable schema, consider modeling it via regular .NET types instead.

**Comment by aleksvujic** [comment]
@roji You wrote in [this comment](https://github.com/dotnet/efcore/issues/29825#issuecomment-2141655659) that mappings like `Dictionary<string, string>` aren't supported yet and suggested using **value converters** instead.

I configured serialization and deserialization with `HasConversion` like this:

```cs
modelBuilder.Entity<Result>(entity =>
{
    entity.OwnsOne(x => x.OutputJSON, x =>
    {
        x.ToJson("output");

        x.Property(y => y.Data)
            .HasConversion(
                v => JsonSerializer.Serialize(v, (JsonSerializerOptions?)null),
                v => JsonSerializer.Deserialize<Dictionary<string, string?>>(v, (JsonSerializerOptions?)null)
            );
    });
});
```

However, when I perform `fitDbContext.Results.ToList()`, the following exception is thrown:

```
System.InvalidOperationException: 'Cannot get the value of a token type 'StartObject' as a string.'
```

I even tried to perform dummy serialization and deserialization like this, but the exception is still the same:

```cs
x.Property(y => y.Data)
    .HasConversion(
        v => "{}",
        v => new Dictionary<string, string?>()
    );
```

Am I using value conversion wrong?

**Comment by roji** [comment]
@aleksvujic I'm a bit confused by the above - above you have an entity type Result, containing a JSON-mapped owned entity (OutputJSON), and **inside that** you want to have a `Data` property that's another JSON sub-document mapped to a dictionary? So a strongly-typed JSON document (OutputJson) containing a weakler-typed JSON document (Data)? Is that the intention here?

If so, note that the entire weakly-typed dictionary (Data) would be serialized as a single string inside the larger JSON document. So you wouldn't get `{ "Data": { "x" : 3 } }`, but rather `{ "Data": "{\"x\": 3 }" }` (note that quotes around Data value. I'm not sure that's what you want.

**Comment by aleksvujic** [comment]
@roji I think you summarized it correctly. JSON in the `output` column in PostgreSQL table looks like this:

```json
{
  "data": {
    "john": "doe",
    "hello": "world"
  }
}
```

The `data` field in the JSON is static and is always there, its _content_ is dynamic (dictionary - `(string, string)` pairs).

That's why I wanted to model `data` in C# like this with `Dictionary<string, string>`:

```cs
// scaffolded DB model
[Table("results", Schema = "fit3")]
public partial class Result
{
    [Key]
    [Column("id")]
    public int Id { get; set; }

    [Column("output", TypeName = "jsonb")]
    public string? Output { get; set; }
}

// custom class with additional strongly-typed OutputJSON property
public partial class Result
{
    public Output? OutputJSON { get; set; }
}

// content of the "data" JSON field (dictionary)
public class Output
{
    [JsonPropertyName("data")]
    public Dictionary<string, string?>? Data { get; set; }
}
```

If I understand correctly, this can't work - this exact problem is tracked by this issue. Correct?

If I format the nested JSON document in the PostgreSQL table as string (shown below), then it successfuly maps it to C# `Dictionary<string, string>`.

```json
{"data": "{ \"john\": \"doe\",  \"hello\": \"world\" }"}
```

**Comment by roji** [comment]
> If I understand correctly, this can't work - this exact problem is tracked by this issue. Correct?

No. EF already supports mapping anything via value converters - whether it be a dictionary or something else. This issue tracks allowing mapping a Dictionary to JSON **without** value converters, which would mean that EF manages the conversion and knows what's going on; that would notably mean that you can e.g. use dictionary lookups inside queries, which is something that you can't do when value-converting a value.

Now, there may be an EF bug specifically when nesting your value-converted dictionary within another JSON owned entity; but that wouldn't be related to this issue.

Can you please open a new issue and include a minimal repro for the problem?
